src/sources/kafka.rs

```diff
@@ -374,105 +390,406 @@ impl SourceConfig for KafkaSourceConfig {
+            // when it is dropped, no new messages will be consumed, and the
+            // task will end when it reaches the end of ack_stream
+            let mut finalizer = Some(finalizer);
+
+            let mut status = PartitionConsumerStatus::NormalExit;
+
+            loop {
+                tokio::select!(
+                    _ = &mut end_signal, if finalizer.is_some() => {
+                        finalizer.take();
+                    },
+                    message = messages.next(), if finalizer.is_some() => match message {
+                        None => unreachable!("MessageStream never calls Ready(None)"),
+                        Some(Err(error)) => match error {
+                            rdkafka::error::KafkaError::PartitionEOF(partition) if exit_eof => {
+                                debug!("EOF for partition {}.", partition);
+                                status = PartitionConsumerStatus::PartitionEOF;
+                                finalizer.take();
+                            },
+                            _ => emit!(KafkaReadError { error }),
+                        },
+                        Some(Ok(msg)) => {
+                            emit!(KafkaBytesReceived {
+                                byte_size: msg.payload_len(),
+                                protocol: "tcp",
+                                topic: msg.topic(),
+                                partition: msg.partition(),
+                            });
+                            parse_message(msg, decoder.clone(), conf.keys(), &mut out, acknowledgements, &finalizer, log_namespace).await;
+                        }
+                    },
+
+                    ack = ack_stream.next() => match ack {
+                        Some((status, entry)) => {
+                            if status == BatchStatus::Delivered {
+                                if let Err(error) =  consumer.store_offset(&entry.topic, entry.partition, entry.offset) {
+                                    emit!(KafkaOffsetUpdateError { error });
+                                }
+                            }
+                        }
+                        None if finalizer.is_none() => {
+                            debug!("acknowledgement stream complete for partition {}:{}.", &tp.0, tp.1);
+                            break
+                        }
+                        None => {
+                            debug!("acknowledgement stream empty for {}:{}", &tp.0, tp.1);
+                        }
+                    }
+                )
+            }
+            (tp, status)
+        });
+        (end_tx, handle)
+    }
+
+    pub fn revoke_partition(&mut self, tp: TopicPartition) {
+        self.expect_drain.insert(tp);
+    }
+
+    pub fn observed_last_ack(&mut self, tp: TopicPartition) {
+        self.observed_drain.insert(tp);
+    }
+
+    pub fn is_drain_complete(&self) -> bool {
+        self.observed_drain == self.expect_drain
```

Comment:
I worry that we can enter a state where we block indefinitely (during rebalance) because `observed_drain` contains elements not in `expect_drain`.  As of now, this might only occur if a consumer tasks ends early due to receiving an EOF signal. However, there *could* be other cases in the future. To make this a more robust check, should we change the logic to check if `expect_drain` is a subset of `observed_drain`? This would also mean that in `clear()`, we would clear `expect_drain` and then remove its elements from `observed_drain`, rather than clearing `observed_drain`. What do you think?

Reply:
I don't have any other cases in mind as of now but I prefer to write defensive code to make things easier down the line 🙂